### ISSUE TITLE
Option to not decay bias with additive_weight_decay

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#25]
Currently, `additive_weight_decay` will decay all the parameters. [Jia et al. 2018](https://arxiv.org/abs/1807.11205) show it is beneficial not to decay the bias parameters (and only decay the weights). Many training examples implement this too, such as [Flax's Imagenet example](https://github.com/google/flax/blob/master/examples/imagenet/train.py#L119-L121). 

One way to support this is to add a `decay_bias: bool` argument to `additive_weight_decay`, then within the decay update:

```python
updates = jax.tree_multimap(
    lambda g, p: g + weight_decay * p if p.ndim > 1 or decay_bias else g, 
    updates, params)
```

As far as I'm aware for common NN layers, the bias always has one dimension, so checking if `p.ndim > 1` is sufficient (please correct me if I'm wrong).

I'd be happy to contribute this if it sounds reasonable.

EDIT: I realized batch norm scale parameters also have only one dimension, so this filter would wrongly include those. However, many training pipelines do not regularize batchnorm scale/bias (e.g. the [reference imagenet implementation from mlperf](https://github.com/mlcommons/training/blob/4b0a874e2eb4c4d7c5e9ff299e99c4324d97acd8/image_classification/tensorflow/official/resnet/resnet_run_loop.py#L314-L329)).

**Comment by mtthss** [comment]
This seems great, maybe to be more clear you could name the argument `decay_scalars`

**Comment by n2cholas** [comment]
I don't think `decay_scalars` is the right term, since we want to not decay parameters with only one dimension (i.e. vectors). `decay_one_dim_params: bool` would be accurate, but is a mouthful 😅. Maybe `decay_vectors: bool`? (though that isn't entirely clear).

**Comment by mtthss** [comment]
Maybe we should just allow to provide a mask instead?

The new constructor argument could be an (optional) py-treee of the same structure as params and grads,
that specifies for each element in the tree whether or not to decay weights.

In the doc string we could also show, as an example, creating the mask by checking `p.ndim > 1`

**Comment by n2cholas** [comment]
I like that! What do you think of making it a [wrapper](https://github.com/deepmind/optax/blob/master/optax/_src/wrappers.py) instead so it's more generic? The signature could be:

```python
def masked_apply(
    inner: transform.GradientTransformation,
    mask: PyTree
) -> transform.GradientTransformation:
```

Another possibility is to make an `apply_if` wrapper, which accepts a predicate. So the signature could be:
```python
def apply_if(
    inner: transform.GradientTransformation,
    predicate: Callable[[jnp.array, jnp.array], bool]  # (gradient, param) -> bool
) -> transform.GradientTransformation:
```
And potential usage:
```python
apply_if(additive_weight_decay, lambda grad, param: param.ndim > 1)
```

**Comment by mtthss** [comment]
I think making it a wrapper sounds great.

I prefer your first proposed API as it allows the user more freedom in **how* the mask is constructed.
For instance in many neural networks libraries in JAX the params structure include some form of scoping, and it is common to filter based on the names instead of the shapes alone.

Maybe name it just `mask` instead of `masked_apply`

**Comment by n2cholas** [comment]
Right, that makes sense. I also agree the name `mask` is cleaner. I'll create a PR for this soon!
